beacon_node/beacon_chain/src/schema_change/migration_schema_v7.rs

```diff
@@ -0,0 +1,331 @@
+            .ok_or_else(|| "Justified root not found".to_string())?;
+        roots_by_epoch.insert(epoch, root);
+    }
+    Ok(roots_by_epoch)
+}
+
+/// Applies a mutator to every node in a chain, starting from the node at the given
+/// `head_index` and iterating through ancestors until the `finalized_root` is reached.
+fn apply_to_chain_of_ancestors<F>(
+    finalized_root: Hash256,
+    head_index: usize,
+    fork_choice: &mut ProtoArrayForkChoice,
+    mut node_mutator: F,
+) -> Result<(), String>
+where
+    F: FnMut(usize, &mut ProtoNode) -> Result<(), String>,
+{
+    let head = fork_choice
+        .core_proto_array_mut()
+        .nodes
+        .get_mut(head_index)
+        .ok_or_else(|| "Head index not found in proto nodes".to_string())?;
+
+    node_mutator(head_index, head)?;
+
+    let mut parent_index_opt = head.parent;
+    let mut parent_opt =
+        parent_index_opt.and_then(|index| fork_choice.core_proto_array_mut().nodes.get_mut(index));
+
+    // Iterate backwards through all parents until there is no reference to a parent or we reach
+    // the `finalized_root` node.
+    while let (Some(parent), Some(parent_index)) = (parent_opt, parent_index_opt) {
+        node_mutator(parent_index, parent)?;
+
+        // Break out of this while loop *after* the `node_mutator` has been applied to the finalized
+        // node.
+        if parent.root == finalized_root {
+            break;
+        }
+
+        // Update parent values
+        parent_index_opt = parent.parent;
+        parent_opt = parent_index_opt
+            .and_then(|index| fork_choice.core_proto_array_mut().nodes.get_mut(index));
+    }
+    Ok(())
+}
+
+/// Finds all heads by finding all nodes in the proto array that are not referenced as parents. Then
+/// checks that these nodes are descendants of the finalized root in order to determine if they are
+/// relevant.
+fn find_finalized_descendant_heads(
+    finalized_root: Hash256,
+    fork_choice: &ProtoArrayForkChoice,
+) -> Vec<HeadInfo> {
+    let nodes_referenced_as_parents: Vec<usize> = fork_choice
+        .core_proto_array()
+        .nodes
+        .iter()
+        .filter_map(|node| node.parent)
+        .collect::<Vec<_>>();
+
+    fork_choice
+        .core_proto_array()
+        .nodes
+        .iter()
+        .enumerate()
+        .filter_map(|(index, node)| {
+            (!nodes_referenced_as_parents.contains(&index)
+                && fork_choice.is_descendant(finalized_root, node.root))
+            .then(|| HeadInfo {
+                index,
+                root: node.root,
+                slot: node.slot,
+            })
+        })
+        .collect::<Vec<_>>()
+}
+
+fn update_store_justified_checkpoint(
+    persisted_fork_choice: &mut PersistedForkChoiceV7,
+    fork_choice: &mut ProtoArrayForkChoice,
+) -> Result<(), String> {
+    let justified_checkpoint = fork_choice
+        .core_proto_array()
+        .nodes
+        .iter()
+        .find_map(|node| {
```

Comment:
In the case where there are multiple, distinct `justified_checkpoints` with a matching `finalized_checkpoint`, using a (forward) `iter` with a `find_map` is likely to discover the *lowest* `justified_checkpoint`.

Perhaps it would be ideal to replace this `find_map` with a `filter_map` and then reduce that set to a single element via [`Iterator::max_by_key`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.max_by_key). That way, we can find the `justified_checkpoint` with the *highest* epoch, which is the value we'd typically expect to see in the store.

Reply:
True, nice catch!